builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java

```diff
@@ -83,13 +92,55 @@ private boolean isSystemStateRequest(String requestURI) {
         return requestURI.endsWith("/api/system/state");
     }
 
-    private boolean isInternalRequest(String authHeader) {
-        return "Internal".equals(authHeader);
+    /**
+     * @param requestURI
+     * @return true if request is made against endpoint which provides keycloak
+     *         status (enabled / disabled), false otherwise
+     */
+    private boolean isKeycloakSettingsRequest(String requestURI) {
+        return requestURI.endsWith("/keycloak/settings");
+    }
+
+    /**
+     * @param authHeader
+     * @return true if "Authorization" header has valid openshift token that is
+     *         used for communication between wsagent and wsmaster, false
+     *         otherwise
+     */
+    private boolean isWorkspaceAgentRequest(String authHeader) {
+        if (authHeader != null && authHeader.startsWith("Wsagent")) {
+            String token = authHeader.replaceFirst("Wsagent ", "");
+            return isValidToken(token);
+        }
+        return false;
     }
 
     private boolean isWebsocketRequest(String requestURI, String requestScheme) {
         return requestURI.endsWith("/ws") || requestURI.endsWith("/eventbus") || requestScheme.equals("ws")
                 || requestScheme.equals("wss") || requestURI.contains("/websocket/")
                 || requestURI.endsWith("/token/user");
     }
+
+    /**
+     * OpenShift default service account token is used in "Authorization" header
+     * for communication between wsagent and wsmaster. The method checks if the
+     * token is valid by fetching namespace info and comparing it with expected 
+     * service account namespace
+     * 
+     * @param token
+     * @return true if openshift token is valid and matches service account namespace, false otherwise
+     */
+    private boolean isValidToken(final String token) {
+        LOG.debug("Validating workspace agent token");
+        Config config = new ConfigBuilder().withOauthToken(token).build();
+        try (OpenShiftClient client = new DefaultOpenShiftClient(config)) {
+            String namespace = client.getConfiguration().getNamespace();
+            LOG.debug("Validating the token against namespace '{}'", namespace);
+            return serviceAccount.getNamespace().equals(namespace);
```

Comment:
I'm not sure I understand the need for ServiceAccount and ServiceAccountTokenParser. Is it possible to use the block 

```java
    try (OpenShiftClient client = new DefaultOpenShiftClient()) {
        this.namespace = client.getNamespace();
    }
    //...
    return this.namespace.equals(tokenNamespace);
```
as a (maybe static block) in `KeycloakAuthenticationFilter`? I understand that the SA token returned is going to be different between wsagent and wsmaster, since they are running in separate pods (and so the token passed to `isValidToken()` won't match), but the namespace should be the same. The above block would get the namespace of the current pod where we're running, and line 137 above would get the namespace of the token passed in. This *should* still avoid the issue of me accessing your workspace, since my token would lead to a different namespace.

I haven't tested it though, so I may have missed something. I've been burned by OS before.